File: 0076-minimum-window-substring/0076-minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    
    if (t.length > s.length) return "";
    
    let tCounts = {};
    
    for (const c of t) {
        tCounts[c] = (tCounts[c] || 0) + 1;
    }
    
    let need = Object.keys(tCounts).length;
    
    for (const c of s.slice(0, t.length)) {
        if (c in tCounts) {
            tCounts[c] -= 1;
            
            if (tCounts[c] === 0) {
                need -= 1;
            }
        }
    }
    
    if (need === 0) return s.slice(0, t.length);
    
    let minSubstring = "";
    let minLen = Infinity;
    
    let left = 0;
    let right = t.length;
    
    while (right < s.length) {
        if (s[right] in tCounts) {
            tCounts[s[right]] -= 1;
            
            if (tCounts[s[right]] === 0) {
                need -= 1;
            }
        }
        
        
        while (need === 0 && left <= s.length - t.length) {
            if (right - left + 1 < minLen) {
                minSubstring = s.slice(left, right + 1);
                minLen = minSubstring.length;
            }
            
            if (s[left] in tCounts) {
                tCounts[s[left]] += 1;
                
                if (tCounts[s[left]] === 1) {
                    need += 1;
                }
            }
            
            left += 1;
        }
        
        right += 1;
    }
    
    return minSubstring;
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (t.isEmpty()) return "";

        Map<Character, Integer> tCounts = new HashMap<>();
        Map<Character, Integer> windowCounts = new HashMap<>();

        for (char c : t.toCharArray()) {
            tCounts.put(c, tCounts.getOrDefault(c, 0) + 1);
        }

        int have = 0;
        int need = tCounts.size();
        int l = 0;
        int[] boundaries = {-1, -1};
        int shortest = Integer.MAX_VALUE;

        for (int r = 0; r < s.length(); r++) {
            char currChar = s.charAt(r);
            windowCounts.put(currChar, windowCounts.getOrDefault(currChar, 0) + 1);

            if (tCounts.containsKey(currChar) && windowCounts.get(currChar).equals(tCounts.get(currChar))) {
                have++;
            }

            while (have == need) {
                if (r - l + 1 < shortest) {
                    shortest = r - l + 1;
                    boundaries[0] = l;
                    boundaries[1] = r;
                }

                char leftChar = s.charAt(l);
                windowCounts.put(leftChar, windowCounts.get(leftChar)-1);

                if (tCounts.containsKey(leftChar) && windowCounts.get(leftChar) < tCounts.get(leftChar)) {
                    have--;
                }
                l++;
            }
        }

        return shortest == Integer.MAX_VALUE ? "" : s.substring(boundaries[0], boundaries[1]+1);
    }
}